ddtrace/internal/datadog/profiling/dd_wrapper/src/code_provenance.cpp

```diff
@@ -0,0 +1,178 @@
+#include "code_provenance.hpp"
+
+#include <memory>
+#include <mutex>
+#include <set>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <vector>
+
+namespace Datadog {
+
+void
+Datadog::CodeProvenance::postfork_child()
+{
+    get_instance().mtx.~mutex();            // Destroy the mutex
+    new (&get_instance().mtx) std::mutex(); // Recreate the mutex
+    get_instance().reset();                 // Reset the state
+}
+
+void
+Datadog::CodeProvenance::set_enabled(bool enable)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    enabled = enable;
+}
+
+bool
+Datadog::CodeProvenance::is_enabled()
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    return enabled;
+}
+
+void
+Datadog::CodeProvenance::set_runtime_version(std::string_view _runtime_version)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    this->runtime_version = _runtime_version;
+}
+
+void
+Datadog::CodeProvenance::set_stdlib_path(std::string_view _stdlib_path)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    this->stdlib_path = _stdlib_path;
+}
+
+void
+CodeProvenance::add_packages(std::unordered_map<std::string_view, std::string_view> distributions)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+
+    if (!enabled) {
+        return;
+    }
+
+    for (const auto& [package_name, version] : distributions) {
+        auto it = packages.find(package_name);
+        if (it == packages.end()) {
+            add_new_package(package_name, version);
+        }
+    }
+}
+
+void
+CodeProvenance::add_filename(std::string_view filename)
+{
+    std::lock_guard<std::mutex> lock(mtx);
+
+    if (!enabled) {
+        return;
+    }
+
+    std::string_view package_name = get_package_name(filename);
+    if (package_name.empty() or package_name == STDLIB) {
+        return;
+    }
+
+    auto it = packages.find(package_name);
+    if (it == packages.end()) {
+        return;
+    }
+
+    const Package* package = it->second.get();
+    if (package) {
+        if (packages_to_files.find(package) == packages_to_files.end()) {
+            packages_to_files[package] = std::set<std::string>();
+        }
+        packages_to_files[package].insert(std::string(filename));
+    }
+}
+
+std::string
+CodeProvenance::serialize_to_json_str()
+{
+    std::lock_guard<std::mutex> lock(mtx);
+
+    if (!enabled) {
+        return "";
+    }
+
+    std::ostringstream out;
+
+    // TODO(taegyunkim): Use a JSON library to serialize this
```

Comment:
https://github.com/nlohmann/json is quite easy to use (not that there's any shortage of good json libraries!)

Reply:
I'm also OK with leaving it like this for now.  No need to overcomplicate our dependencies for one function.  :)